plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/ProductQueryBuilder.php

```diff
@@ -0,0 +1,1186 @@
+<?php
+declare(strict_types=1);
+
+namespace Automattic\WooCommerce\Blocks\BlockTypes\ProductCollection;
+
+use Automattic\WooCommerce\Blocks\BlockTypes\AttributeFilter;
+use Automattic\WooCommerce\Blocks\BlockTypes\PriceFilter;
+use Automattic\WooCommerce\Blocks\BlockTypes\RatingFilter;
+use Automattic\WooCommerce\Blocks\BlockTypes\StockFilter;
+use WP_Query;
+use WC_Tax;
+
+/**
+ * ProductQueryBuilder class.
+ * Responsible for constructing and modifying product queries.
+ */
+class ProductQueryBuilder {
+
+	/**
+	 * All query args from WP_Query.
+	 *
+	 * @var array
+	 */
+	protected $valid_query_vars;
+
+	/**
+	 * Orderby options not natively supported by WordPress REST API
+	 *
+	 * @var array
+	 */
+	protected $custom_order_opts = array( 'popularity', 'rating', 'post__in', 'price', 'sales', 'menu_order' );
+
+	/**
+	 * All the query args related to the filter by attributes block.
+	 *
+	 * @var array
+	 */
+	protected $attributes_filter_query_args = array();
+
+	/**
+	 * Collection handler store.
+	 *
+	 * @var array
+	 */
+	protected $collection_handler_store = array();
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->valid_query_vars = $this->get_valid_query_vars();
+		add_filter( 'posts_clauses', array( $this, 'add_price_range_filter_posts_clauses' ), 10, 2 );
+	}
+
+	/**
+	 * Set collection handler store.
+	 *
+	 * @param string $collection_name The name of the custom collection.
+	 * @param array  $handlers        Collection handlers.
+	 */
+	public function set_collection_handler_store( $collection_name, $handlers ) {
+		$this->collection_handler_store[ $collection_name ] = $handlers;
+	}
```

Comment:
We will need this setter to set the `collection_handler_store` at `ProductCollectionController` level. This will change with the introduction of the [class CollectionHandlerRegistry](https://github.com/woocommerce/woocommerce/issues/52417).